cardano-api/src/Cardano/Api/Governance/Poll.hs

```diff
@@ -0,0 +1,510 @@
+               <$> fmap VKey (DSIGN.rawDeserialiseVerKeyDSIGN vk)
+               <*> fmap DSIGN.SignedDSIGN (DSIGN.rawDeserialiseSigDSIGN sig)
+           Just _  ->
+             Left $ malformedField (fieldPath lbl k) "List"
+           Nothing ->
+             orElse
+
+signPollAnswerWith
+  :: GovernancePollAnswer
+  -> Either (SignKeyVRF StandardCrypto) (SignKeyDSIGN StandardCrypto)
+  -> GovernancePollWitness
+signPollAnswerWith answer =
+  either
+    (\sk -> GovernancePollWitnessVRF
+      (VRF.deriveVerKeyVRF sk)
+      (snd $ VRF.evalVRF () answer sk)
+    )
+    (\sk -> GovernancePollWitnessColdKey
+      (VKey (DSIGN.deriveVerKeyDSIGN sk))
+      (signedDSIGN @StandardCrypto sk answer)
+    )
+
+-- ----------------------------------------------------------------------------
+-- Governance Poll Verification
+--
+
+data GovernancePollError
+  = ErrGovernancePollMismatch
+  | ErrGovernancePollInvalidAnswer GovernancePollInvalidAnswerError
+  | ErrGovernancePollInvalidWitness
+  deriving Show
+
+data GovernancePollInvalidAnswerError = GovernancePollInvalidAnswerError
+  { invalidAnswerAcceptableAnswers :: [(Word, Text)]
+  , invalidAnswerReceivedAnswer :: Word
+  }
+  deriving Show
+
+renderGovernancePollError :: GovernancePollError -> Text
+renderGovernancePollError err =
+  case err of
+    ErrGovernancePollMismatch ->
+      "Answer's poll doesn't match provided poll (hash mismatch)."
+    ErrGovernancePollInvalidAnswer invalidAnswer ->
+        mconcat
+          [ "Invalid answer ("
+          , textShow (invalidAnswerReceivedAnswer invalidAnswer)
+          , ") not part of the poll."
+          , "\n"
+          , "Accepted answers:"
+          , "\n"
+          , Text.intercalate "\n"
+              [ mconcat
+                  [ textShow ix
+                  , " → "
+                  , answer
+                  ]
+              | (ix, answer) <- invalidAnswerAcceptableAnswers invalidAnswer
+              ]
+          ]
+    ErrGovernancePollInvalidWitness ->
+      "Invalid witness for the answer: the proof / signature doesn't hold."
+
+verifyPollAnswer
+  :: GovernancePoll
+  -> GovernancePollAnswer
+  -> GovernancePollWitness
+  -> Either GovernancePollError ()
+verifyPollAnswer poll answer witness = do
+  when (hashGovernancePoll poll /= govAnsPoll answer) $
+    Left ErrGovernancePollMismatch
+
+  when (govAnsChoice answer >= fromIntegral (length (govPollAnswers poll))) $ do
+    let invalidAnswerReceivedAnswer = govAnsChoice answer
+    let invalidAnswerAcceptableAnswers = zip [0..] (govPollAnswers poll)
+    Left $ ErrGovernancePollInvalidAnswer $ GovernancePollInvalidAnswerError
+      { invalidAnswerReceivedAnswer
+      , invalidAnswerAcceptableAnswers
+      }
+
+  unless isValid $
+    Left ErrGovernancePollInvalidWitness
+ where
+   isValid =
+    case witness of
+      GovernancePollWitnessVRF vk proof ->
+        VRF.verifyVRF () vk answer (undefined, proof)
```

Comment:
This is fishy. It's at minimum being unreasonably cozy with the implementation, rather than relying on the class interface.

Reply:
Yes this is a slight abuse of the interface because the current class API requires the `Output` (the `undefined` term here) without making any use of it since it is actually irrelevant to the verification process. Since in this scenario we do not even care about the outputs (as you're rightfully pointing out in the CIP's comment), I didn't feel like bothering end-users with them either just for the sake of satisfying an interface that's arguably not restrictive enough.